core/src/main/java/org/elasticsearch/action/ingest/IngestActionFilter.java

```diff
@@ -57,25 +58,25 @@ public IngestActionFilter(Settings settings, IngestBootstrapper bootstrapper) {
 
     @Override
     public void apply(Task task, String action, ActionRequest request, ActionListener listener, ActionFilterChain chain) {
-        String pipelineId = request.getFromContext(PIPELINE_ID_PARAM_CONTEXT_KEY);
-        if (pipelineId == null) {
-            pipelineId = request.getHeader(PIPELINE_ID_PARAM);
-            if (pipelineId == null) {
-                chain.proceed(task, action, request, listener);
+
+        if (IndexAction.NAME.equals(action)) {
+            assert request instanceof IndexRequest;
+            IndexRequest indexRequest = (IndexRequest) request;
+            if (Strings.hasText(indexRequest.pipeline())) {
+                processIndexRequest(task, action, listener, chain, (IndexRequest) request);
                 return;
             }
         }
-
-        if (request instanceof IndexRequest) {
-            processIndexRequest(task, action, listener, chain, (IndexRequest) request, pipelineId);
-        } else if (request instanceof BulkRequest) {
+        if (BulkAction.NAME.equals(action)) {
+            assert request instanceof BulkRequest;
```

Comment:
I think we need check whether pipeline has be specified on the bulk request or on any sub requests here. Right now we always delegate to the pipeline execution service that will execute on the ingest thread pool, which we should prevent if no pipeline has been specified.